rustls/src/server/tls12.rs

```diff
@@ -63,25 +63,26 @@ impl hs::State for ExpectCertificate {
                 TLSError::General("client rejected by client_auth_mandatory".into())
             })?;
 
-        if cert_chain.is_empty() {
-            if !mandatory {
-                debug!("client auth requested but no certificate supplied");
-                self.handshake
-                    .transcript
-                    .abandon_client_auth();
-                return Ok(self.into_expect_tls12_client_kx(None));
-            }
-            sess.common
-                .send_fatal_alert(AlertDescription::CertificateRequired);
-            return Err(TLSError::NoCertificatesPresented);
-        }
-
         trace!("certs {:?}", cert_chain);
 
+        let (end_entity, intermediates) = match cert_chain.split_first() {
+            None => {
+                if !mandatory {
+                    debug!("client auth requested but no certificate supplied");
+                    self.handshake.transcript.abandon_client_auth();
+                    return Ok(self.into_expect_tls12_client_kx(None));
+                }
+                sess.common
+                    .send_fatal_alert(AlertDescription::CertificateRequired);
+                return Err(TLSError::NoCertificatesPresented);
+            }
```

Comment:
Maybe write it like this? Seems slightly easier to follow.

```suggestion
            None if !mandatory => {
                    debug!("client auth requested but no certificate supplied");
                    self.handshake.transcript.abandon_client_auth();
                    return Ok(self.into_expect_tls12_client_kx(None));
            }
            None => {
                sess.common
                    .send_fatal_alert(AlertDescription::CertificateRequired);
                return Err(TLSError::NoCertificatesPresented);
            }
```

Reply:
I'd rather keep the current structure for now, because it is intentionally the same structure that was used in the code previously. Maybe we can make that kind of refactoring in a separate commit. I want to keep the focus of this commit on eliminating the use of the array indexing and improving the public API.

Reply:
Fair enough.